app/src/main/java/com/amaze/filemanager/filesystem/operations/CopyOperation.java

```diff
@@ -0,0 +1,266 @@
+import com.amaze.filemanager.filesystem.RootHelper;
+import com.amaze.filemanager.filesystem.operations.exceptions.CancellationIOException;
+import com.amaze.filemanager.filesystem.operations.exceptions.ShellNotRunningIOException;
+import com.amaze.filemanager.filesystem.operations.singlefile.DeleteOperation;
+import com.amaze.filemanager.utils.OpenMode;
+import com.amaze.filemanager.utils.ProgressHandler;
+import com.amaze.filemanager.utils.RootUtils;
+import com.amaze.filemanager.utils.files.FileUtils;
+import com.amaze.filemanager.utils.files.GenericCopyUtil;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import androidx.annotation.Nullable;
+
+public class CopyOperation extends AbstractOperation {
+	private final CopyService copyService;
+	private final Context context;
+	private final ServiceWatcherUtil watcherUtil;
+	private final ProgressHandler progressHandler;
+	private final List<HybridFileParcelable> sourceFiles;
+	private final String targetPath;
+	private final boolean move;
+	private final OpenMode mode;
+	private final boolean isRootExplorer;
+
+	private int sourceProgress = 0;
+	private ArrayList<HybridFile> failedFOps;
+
+	/**
+	 * @param mode target file open mode (current path's open mode)
+	 */
+	public CopyOperation(CopyService copyService, ServiceWatcherUtil watcherUtil,
+	                     ProgressHandler progressHandler,
+	                     List<HybridFileParcelable> sourceFiles, String targetPath,
+	                     boolean move, OpenMode mode, boolean isRootExplorer) {
+		this.copyService = copyService;
+		this.context = copyService;
+		this.watcherUtil = watcherUtil;
+		this.progressHandler = progressHandler;
+		this.sourceFiles = sourceFiles;
+		this.targetPath = targetPath;
+		this.move = move;
+		this.mode = mode;
+		this.isRootExplorer = isRootExplorer;
+
+		failedFOps = new ArrayList<>();
+	}
+
+
+	@Override
+	protected boolean check() {
+		return true;
+	}
+
+	@Override
+	protected void operate() throws IOException {
+		// initial start of copy, initiate the watcher
+		watcherUtil.watch(copyService);
+
+		if (FileUtil.checkFolder((targetPath), context) == 1) {
+			for (int i = 0; i < sourceFiles.size(); i++) {
+				sourceProgress = i;
+				HybridFileParcelable f1 = (sourceFiles.get(i));
+
+				try {
+					HybridFile hFile;
+					if (targetPath.contains(context.getExternalCacheDir().getPath())) {
+						// the target open mode is not the one we're currently in!
+						// we're processing the file for cache
+						hFile = new HybridFile(OpenMode.FILE, targetPath, sourceFiles.get(i).getName(),
+								f1.isDirectory());
+					} else {
+
+						// the target open mode is where we're currently at
+						hFile = new HybridFile(mode, targetPath, sourceFiles.get(i).getName(),
+								f1.isDirectory());
+					}
+
+					if (progressHandler.getCancelled()) {
+						throw new CancellationIOException();
+					}
```

Comment:
Throwing an exception in a `try` block is generally considered a bad practice. I think we can refactor this out.

Reply:
How?